Fix tests failures in subquery_test.py

```
FAILED ../../../../integration_tests/src/main/python/subquery_test.py::test_scalar_subquery_struct
FAILED ../../../../integration_tests/src/main/python/subquery_test.py::test_scalar_subquery_struct
FAILED ../../../../integration_tests/src/main/python/subquery_test.py::test_scalar_subquery_struct
FAILED ../../../../integration_tests/src/main/python/subquery_test.py::test_scalar_subquery_array
FAILED ../../../../integration_tests/src/main/python/subquery_test.py::test_scalar_subquery_struct
FAILED ../../../../integration_tests/src/main/python/subquery_test.py::test_scalar_subquery_array
FAILED ../../../../integration_tests/src/main/python/subquery_test.py::test_scalar_subquery_array
FAILED ../../../../integration_tests/src/main/python/subquery_test.py::test_scalar_subquery_struct
FAILED ../../../../integration_tests/src/main/python/subquery_test.py::test_scalar_subquery_struct
FAILED ../../../../integration_tests/src/main/python/subquery_test.py::test_scalar_subquery_struct
FAILED ../../../../integration_tests/src/main/python/subquery_test.py::test_scalar_subquery_struct
FAILED ../../../../integration_tests/src/main/python/subquery_test.py::test_scalar_subquery_array
FAILED ../../../../integration_tests/src/main/python/subquery_test.py::test_scalar_subquery_array
FAILED ../../../../integration_tests/src/main/python/subquery_test.py::test_scalar_subquery_struct
FAILED ../../../../integration_tests/src/main/python/subquery_test.py::test_scalar_subquery_array
FAILED ../../../../integration_tests/src/main/python/subquery_test.py::test_scalar_subquery_array
FAILED ../../../../integration_tests/src/main/python/subquery_test.py::test_scalar_subquery_struct
FAILED ../../../../integration_tests/src/main/python/subquery_test.py::test_scalar_subquery_struct
FAILED ../../../../integration_tests/src/main/python/subquery_test.py::test_scalar_subquery_array
FAILED ../../../../integration_tests/src/main/python/subquery_test.py::test_scalar_subquery_array
FAILED ../../../../integration_tests/src/main/python/subquery_test.py::test_scalar_subquery_map
```

More info: `subquery_test.py::test_scalar_subquery_struct` fails because of a possible bug in `libcudf`:
```
E                       Suppressed: com.nvidia.spark.rapids.jni.GpuRetryOOM: injected RetryOOM
E                               at ai.rapids.cudf.ColumnView.hasNonEmptyNulls(Native Method)
E                               at ai.rapids.cudf.ColumnView.hasNonEmptyNulls(ColumnView.java:5218)
E                               at ai.rapids.cudf.AssertEmptyNulls.assertNullsAreEmpty(AssertEmptyNulls.java:33)
E                               at ai.rapids.cudf.ColumnView.<init>(ColumnView.java:58)
...
```
Somewhere along the way, it appears that the output columns have non-empty null rows.  This will need to be fixed in `libcudf`.  I'll raise a separate bug to track this.

--- a/integration_tests/src/main/python/subquery_test.py
+++ b/integration_tests/src/main/python/subquery_test.py
@@ -1,4 +1,4 @@
-# Copyright (c) 2021-2022, NVIDIA CORPORATION.
+# Copyright (c) 2021-2024, NVIDIA CORPORATION.
 #
 # Licensed under the Apache License, Version 2.0 (the "License");
 # you may not use this file except in compliance with the License.

--- a/integration_tests/src/main/python/subquery_test.py
+++ b/integration_tests/src/main/python/subquery_test.py
@@ -13,46 +13,51 @@
 # limitations under the License.
 
 import pytest
-from asserts import assert_gpu_and_cpu_are_equal_collect, assert_gpu_and_cpu_are_equal_sql, assert_gpu_and_cpu_error
+from asserts import assert_gpu_and_cpu_are_equal_sql, assert_gpu_and_cpu_error
+from conftest import spark_tmp_table_factory
 from data_gen import *
 from marks import *
 
 @ignore_order(local=True)
 @pytest.mark.parametrize('data_gen', all_basic_gens, ids=idfn)
-def test_scalar_subquery_basics(data_gen):
+def test_scalar_subquery_basics(spark_tmp_table_factory, data_gen):
     # Fix num_slices at 1 to make sure that first/last returns same results under CPU and GPU.
+    table_name = spark_tmp_table_factory.get()
     assert_gpu_and_cpu_are_equal_sql(
         lambda spark: gen_df(spark, [('a', data_gen)], num_slices=1),
-        'table',
-        '''select a, (select last(a) from table)
-        from table
-        where a > (select first(a) from table)
-        ''')
+        table_name,
+        '''select a, (select last(a) from {})
+        from {}
+        where a > (select first(a) from {})
+        '''.format(table_name, table_name, table_name))
 
 
 @ignore_order(local=True)
 @pytest.mark.parametrize('basic_gen', all_basic_gens, ids=idfn)
-def test_scalar_subquery_struct(basic_gen):
+def test_scalar_subquery_struct(spark_tmp_table_factory, basic_gen):
     # single-level struct
     gen = [('ss', StructGen([['a', basic_gen], ['b', basic_gen]]))]
+    table_name = spark_tmp_table_factory.get()
     assert_gpu_and_cpu_are_equal_sql(
         # Fix num_slices at 1 to make sure that first/last returns same results under CPU and GPU.
         lambda spark: gen_df(spark, gen, num_slices=1),
-        'table',
-        '''select ss, (select last(ss) from table)
-        from table
-        where (select first(ss) from table).b > ss.a
-        ''')
+        table_name,
+        '''select ss, (select last(ss) from {})
+        from {}
+        where (select first(ss) from {}).b > ss.a
+        '''.format(table_name, table_name, table_name))
+
     # nested struct
     gen = [('ss', StructGen([['child', StructGen([['c0', basic_gen]])]]))]
     assert_gpu_and_cpu_are_equal_sql(
         # Fix num_slices at 1 to make sure that first/last returns same results under CPU and GPU.
         lambda spark: gen_df(spark, gen, num_slices=1),
-        'table',
-        '''select ss, (select last(ss) from table)
-        from table
-        where (select first(ss) from table)['child']['c0'] > ss.child.c0
-        ''')
+        table_name,
+        '''select ss, (select last(ss) from {})
+        from {}
+        where (select first(ss) from {})['child']['c0'] > ss.child.c0
+        '''.format(table_name, table_name, table_name))
+
     # struct of array
     # Note: The test query accesses the first two elements of the array.  The datagen is set up
     #       to generate arrays of a minimum of two elements.  Otherwise, the test will fail in ANSI mode.
@@ -62,17 +67,17 @@ def test_scalar_subquery_struct(basic_gen):
     assert_gpu_and_cpu_are_equal_sql(
         # Fix num_slices at 1 to make sure that first/last returns same results under CPU and GPU.
         lambda spark: gen_df(spark, gen, length=100, num_slices=1),
-        'table',
-        '''select sort_array(ss.arr), sort_array((select last(ss) from table)['arr'])
-        from table
-        where (select first(ss) from table).arr[0] > ss.arr[1]
-        ''')
+        table_name,
+        '''select sort_array(ss.arr), sort_array((select last(ss) from {})['arr'])
+        from {}
+        where (select first(ss) from {}).arr[0] > ss.arr[1]
+        '''.format(table_name, table_name, table_name))
 
 
 @ignore_order(local=True)
 @pytest.mark.parametrize('is_ansi_enabled', [False, True])
 @pytest.mark.parametrize('basic_gen', all_basic_gens, ids=idfn)
-def test_scalar_subquery_array(is_ansi_enabled, basic_gen):
+def test_scalar_subquery_array(spark_tmp_table_factory, is_ansi_enabled, basic_gen):
     """
     For this test, all the array inputs are sized so that ArrayIndexOutOfBounds conditions are
     avoided.  This is to ensure that the tests don't fail with exceptions in ANSI mode.
@@ -80,18 +85,19 @@ def test_scalar_subquery_array(is_ansi_enabled, basic_gen):
     already tested as part of array_test.py::test_array_item_ansi_fail_invalid_index.
     """
     conf = {'spark.sql.ansi.enabled': is_ansi_enabled}
+    table_name = spark_tmp_table_factory.get()
 
     # single-level array
     test_array_gen = ArrayGen(basic_gen, min_length=1 if is_ansi_enabled else 0)
     assert_gpu_and_cpu_are_equal_sql(
         # Fix num_slices at 1 to make sure that first/last returns same results under CPU and GPU.
         lambda spark: gen_df(spark, [('arr', test_array_gen)], num_slices=1),
-        'my_array_table',
+        table_name,
         '''select sort_array(arr),
-                  sort_array((select last(arr) from my_array_table))
-        from my_array_table
-        where (select first(arr) from my_array_table)[0] > arr[0]
-        ''',
+                  sort_array((select last(arr) from {}))
+        from {}
+        where (select first(arr) from {})[0] > arr[0]
+        '''.format(table_name, table_name, table_name),
         conf=conf)
 
     # nested array
@@ -100,12 +106,12 @@ def test_scalar_subquery_array(is_ansi_enabled, basic_gen):
     assert_gpu_and_cpu_are_equal_sql(
         # Fix num_slices at 1 to make sure that first/last returns same results under CPU and GPU.
         lambda spark: gen_df(spark, [('arr', test_array_gen)], length=100, num_slices=1),
-        'table',
+        table_name,
         '''select sort_array(arr[10]),
-                  sort_array((select last(arr) from table)[10])
-        from table
-        where (select first(arr) from table)[0][1] > arr[0][1]
-        ''',
+                  sort_array((select last(arr) from {})[10])
+        from {}
+        where (select first(arr) from {})[0][1] > arr[0][1]
+        '''.format(table_name, table_name, table_name),
         conf=conf)
 
     # array of struct
@@ -113,16 +119,16 @@ def test_scalar_subquery_array(is_ansi_enabled, basic_gen):
     assert_gpu_and_cpu_are_equal_sql(
         # Fix num_slices at 1 to make sure that first/last returns same results under CPU and GPU.
         lambda spark: gen_df(spark, [('arr', test_array_gen)], length=100, num_slices=1),
-        'table',
-        '''select arr[10].a, (select last(arr) from table)[10].a
-        from table
-        where (select first(arr) from table)[0].a > arr[0].a
-        ''',
+        table_name,
+        '''select arr[10].a, (select last(arr) from {})[10].a
+        from {}
+        where (select first(arr) from {})[0].a > arr[0].a
+        '''.format(table_name, table_name, table_name),
         conf=conf)
 
 
 @ignore_order(local=True)
-def test_scalar_subquery_array_ansi_mode_failures():
+def test_scalar_subquery_array_ansi_mode_failures(spark_tmp_table_factory):
     """
     This tests the case where the array scalar returned from a subquery might be indexed into
     with an out-of-range index value. With ANSI mode enabled, an exception is expected.
@@ -130,9 +136,9 @@ def test_scalar_subquery_array_ansi_mode_failures():
     A more thorough test for invalid indices is done in array_test.py::test_array_item_ansi_fail_invalid_index,
     and is out of the scope of this test.
     """
+    table_name = spark_tmp_table_factory.get()
 
     def test_function(spark):
-        table_name = 'scalar_subquery_array_input'
         # Fix num_slices at 1 to make sure that first/last returns same results under CPU and GPU.
         df = gen_df(spark, [('arr', ArrayGen(long_gen))], num_slices=1)
         df.createOrReplaceTempView(table_name)
@@ -151,38 +157,41 @@ def test_scalar_subquery_array_ansi_mode_failures():
 
 
 @ignore_order(local=True)
-def test_scalar_subquery_map():
+def test_scalar_subquery_map(spark_tmp_table_factory):
     # Note: For this test, all the array inputs are sized so that ArrayIndexOutOfBounds conditions are
     #       avoided.  This is to ensure that the tests don't fail with exceptions in ANSI mode.
     #       Note that no meaningful test coverage is lost here.  ArrayIndexOutOfBounds exceptions are
     #       already tested as part of array_test.py::test_array_item_ansi_fail_invalid_index.
+    table_name = spark_tmp_table_factory.get()
     map_gen = map_string_string_gen[0]
     assert_gpu_and_cpu_are_equal_sql(
         # Fix num_slices at 1 to make sure that first/last returns same results under CPU and GPU.
         lambda spark: gen_df(spark, [('kv', map_gen)], length=100, num_slices=1),
-        'table',
+        table_name,
         '''select kv['key_0'],
-                  (select first(kv) from table)['key_1'],
-                  (select last(kv) from table)['key_2']
-        from table
-        ''')
+                  (select first(kv) from {})['key_1'],
+                  (select last(kv) from {})['key_2']
+        from {}
+        '''.format(table_name, table_name, table_name))
+
     # array of map
     assert_gpu_and_cpu_are_equal_sql(
         # Fix num_slices at 1 to make sure that first/last returns same results under CPU and GPU.
         lambda spark: gen_df(spark, [('arr', ArrayGen(map_gen, min_length=1))], length=100, num_slices=1),
-        'table',
+        table_name,
         '''select arr[0]['key_0'],
-                  (select first(arr) from table)[0]['key_1'],
-                  (select last(arr[0]) from table)['key_2']
-        from table
-        ''')
+                  (select first(arr) from {})[0]['key_1'],
+                  (select last(arr[0]) from {})['key_2']
+        from {}
+        '''.format(table_name, table_name, table_name))
+
     # struct of map
     assert_gpu_and_cpu_are_equal_sql(
         # Fix num_slices at 1 to make sure that first/last returns same results under CPU and GPU.
         lambda spark: gen_df(spark, [('ss', StructGen([['kv', map_gen]]))], length=100, num_slices=1),
-        'table',
+        table_name,
         '''select ss['kv']['key_0'],
-                  (select first(ss) from table)['kv']['key_1'],
-                  (select last(ss.kv) from table)['key_2']
-        from table
-        ''')
+                  (select first(ss) from {})['kv']['key_1'],
+                  (select last(ss.kv) from {})['key_2']
+        from {}
+        '''.format(table_name, table_name, table_name))

--- a/integration_tests/src/main/python/subquery_test.py
+++ b/integration_tests/src/main/python/subquery_test.py
@@ -70,61 +70,66 @@ def test_scalar_subquery_struct(basic_gen):
 
 
 @ignore_order(local=True)
+@pytest.mark.parametrize('is_ansi_enabled', [False, True])
 @pytest.mark.parametrize('basic_gen', all_basic_gens, ids=idfn)
-def test_scalar_subquery_array(basic_gen):
-    # Note: For this test, all the array inputs are sized so that ArrayIndexOutOfBounds conditions are
-    #       avoided.  This is to ensure that the tests don't fail with exceptions in ANSI mode.
-    #       Note that no meaningful test coverage is lost here.  ArrayIndexOutOfBounds exceptions are
-    #       already tested as part of array_test.py::test_array_item_ansi_fail_invalid_index.
+def test_scalar_subquery_array(is_ansi_enabled, basic_gen):
+    """
+    For this test, all the array inputs are sized so that ArrayIndexOutOfBounds conditions are
+    avoided.  This is to ensure that the tests don't fail with exceptions in ANSI mode.
+    Note that no meaningful test coverage is lost here.  ArrayIndexOutOfBounds exceptions are
+    already tested as part of array_test.py::test_array_item_ansi_fail_invalid_index.
+    """
+    conf = {'spark.sql.ansi.enabled': is_ansi_enabled}
+
     # single-level array
+    test_array_gen = ArrayGen(basic_gen, min_length=1 if is_ansi_enabled else 0)
     assert_gpu_and_cpu_are_equal_sql(
         # Fix num_slices at 1 to make sure that first/last returns same results under CPU and GPU.
-        lambda spark: gen_df(spark, [('arr', ArrayGen(basic_gen, min_length=1))], num_slices=1),
+        lambda spark: gen_df(spark, [('arr', test_array_gen)], num_slices=1),
         'table',
         '''select sort_array(arr),
                   sort_array((select last(arr) from table))
         from table
         where (select first(arr) from table)[0] > arr[0]
-        ''')
+        ''',
+        conf=conf)
+
     # nested array
+    test_array_gen = ArrayGen(ArrayGen(basic_gen, min_length=2 if is_ansi_enabled else 0),
+                              min_length=11 if is_ansi_enabled else 0)
     assert_gpu_and_cpu_are_equal_sql(
         # Fix num_slices at 1 to make sure that first/last returns same results under CPU and GPU.
-        lambda spark: gen_df(spark, [('arr', ArrayGen(ArrayGen(basic_gen, min_length=2), min_length=11))]
-                             , length=100
-                             , num_slices=1),
+        lambda spark: gen_df(spark, [('arr', test_array_gen)], length=100, num_slices=1),
         'table',
         '''select sort_array(arr[10]),
                   sort_array((select last(arr) from table)[10])
         from table
         where (select first(arr) from table)[0][1] > arr[0][1]
-        ''')
+        ''',
+        conf=conf)
+
     # array of struct
+    test_array_gen = ArrayGen(StructGen([['a', basic_gen]]), min_length=11 if is_ansi_enabled else 0)
     assert_gpu_and_cpu_are_equal_sql(
         # Fix num_slices at 1 to make sure that first/last returns same results under CPU and GPU.
-        lambda spark: gen_df(spark, [('arr', ArrayGen(StructGen([['a', basic_gen]]), min_length=11))]
-                             , length=100
-                             , num_slices=1),
+        lambda spark: gen_df(spark, [('arr', test_array_gen)], length=100, num_slices=1),
         'table',
         '''select arr[10].a, (select last(arr) from table)[10].a
         from table
         where (select first(arr) from table)[0].a > arr[0].a
-        ''')
+        ''',
+        conf=conf)
 
 
 @ignore_order(local=True)
-@pytest.mark.parametrize('is_ansi_enabled', [False, True])
-def test_scalar_subquery_array_ansi_mode_failures(is_ansi_enabled):
+def test_scalar_subquery_array_ansi_mode_failures():
     """
     This tests the case where the array scalar returned from a subquery might be indexed into
-    with an out-of-range index value:
-      - With ANSI mode enabled, an exception is expected.
-      - With ANSI mode disabled, a NULL value is expected.
-    This is only to test the corner case that used to be tested in `test_scalar_subquery_array`,
-    before ANSI support was added.
+    with an out-of-range index value. With ANSI mode enabled, an exception is expected.
+
     A more thorough test for invalid indices is done in array_test.py::test_array_item_ansi_fail_invalid_index,
     and is out of the scope of this test.
     """
-    conf = {'spark.sql.ansi.enabled': is_ansi_enabled}
 
     def test_function(spark):
         table_name = 'scalar_subquery_array_input'
@@ -139,13 +144,10 @@ def test_scalar_subquery_array_ansi_mode_failures(is_ansi_enabled):
         '''.format(table_name, table_name, table_name)
         return spark.sql(query)
 
-    if is_ansi_enabled:
-        assert_gpu_and_cpu_error(
-            lambda spark: test_function(spark).collect(),
-            conf=conf,
-            error_message='ArrayIndexOutOfBoundsException')
-    else:
-        assert_gpu_and_cpu_are_equal_collect(test_function, conf=conf)
+    assert_gpu_and_cpu_error(
+        lambda spark: test_function(spark).collect(),
+        conf=ansi_enabled_conf,
+        error_message='ArrayIndexOutOfBoundsException')
 
 
 @ignore_order(local=True)

--- a/integration_tests/src/main/python/subquery_test.py
+++ b/integration_tests/src/main/python/subquery_test.py
@@ -26,10 +26,10 @@ def test_scalar_subquery_basics(spark_tmp_table_factory, data_gen):
     assert_gpu_and_cpu_are_equal_sql(
         lambda spark: gen_df(spark, [('a', data_gen)], num_slices=1),
         table_name,
-        '''select a, (select last(a) from {})
-        from {}
-        where a > (select first(a) from {})
-        '''.format(table_name, table_name, table_name))
+        f'''select a, (select last(a) from {table_name})
+        from {table_name}
+        where a > (select first(a) from {table_name})
+        ''')
 
 
 @ignore_order(local=True)
@@ -42,10 +42,10 @@ def test_scalar_subquery_struct(spark_tmp_table_factory, basic_gen):
         # Fix num_slices at 1 to make sure that first/last returns same results under CPU and GPU.
         lambda spark: gen_df(spark, gen, num_slices=1),
         table_name,
-        '''select ss, (select last(ss) from {})
-        from {}
-        where (select first(ss) from {}).b > ss.a
-        '''.format(table_name, table_name, table_name))
+        f'''select ss, (select last(ss) from {table_name})
+        from {table_name}
+        where (select first(ss) from {table_name}).b > ss.a
+        ''')
 
     # nested struct
     gen = [('ss', StructGen([['child', StructGen([['c0', basic_gen]])]]))]
@@ -53,10 +53,10 @@ def test_scalar_subquery_struct(spark_tmp_table_factory, basic_gen):
         # Fix num_slices at 1 to make sure that first/last returns same results under CPU and GPU.
         lambda spark: gen_df(spark, gen, num_slices=1),
         table_name,
-        '''select ss, (select last(ss) from {})
-        from {}
-        where (select first(ss) from {})['child']['c0'] > ss.child.c0
-        '''.format(table_name, table_name, table_name))
+        f'''select ss, (select last(ss) from {table_name})
+        from {table_name}
+        where (select first(ss) from {table_name})['child']['c0'] > ss.child.c0
+        ''')
 
     # struct of array
     # Note: The test query accesses the first two elements of the array.  The datagen is set up
@@ -68,10 +68,10 @@ def test_scalar_subquery_struct(spark_tmp_table_factory, basic_gen):
         # Fix num_slices at 1 to make sure that first/last returns same results under CPU and GPU.
         lambda spark: gen_df(spark, gen, length=100, num_slices=1),
         table_name,
-        '''select sort_array(ss.arr), sort_array((select last(ss) from {})['arr'])
-        from {}
-        where (select first(ss) from {}).arr[0] > ss.arr[1]
-        '''.format(table_name, table_name, table_name))
+        f'''select sort_array(ss.arr), sort_array((select last(ss) from {table_name})['arr'])
+        from {table_name}
+        where (select first(ss) from {table_name}).arr[0] > ss.arr[1]
+        ''')
 
 
 @ignore_order(local=True)
@@ -93,11 +93,11 @@ def test_scalar_subquery_array(spark_tmp_table_factory, is_ansi_enabled, basic_g
         # Fix num_slices at 1 to make sure that first/last returns same results under CPU and GPU.
         lambda spark: gen_df(spark, [('arr', test_array_gen)], num_slices=1),
         table_name,
-        '''select sort_array(arr),
-                  sort_array((select last(arr) from {}))
-        from {}
-        where (select first(arr) from {})[0] > arr[0]
-        '''.format(table_name, table_name, table_name),
+        f'''select sort_array(arr),
+                  sort_array((select last(arr) from {table_name}))
+        from {table_name}
+        where (select first(arr) from {table_name})[0] > arr[0]
+        ''',
         conf=conf)
 
     # nested array
@@ -107,11 +107,11 @@ def test_scalar_subquery_array(spark_tmp_table_factory, is_ansi_enabled, basic_g
         # Fix num_slices at 1 to make sure that first/last returns same results under CPU and GPU.
         lambda spark: gen_df(spark, [('arr', test_array_gen)], length=100, num_slices=1),
         table_name,
-        '''select sort_array(arr[10]),
-                  sort_array((select last(arr) from {})[10])
-        from {}
-        where (select first(arr) from {})[0][1] > arr[0][1]
-        '''.format(table_name, table_name, table_name),
+        f'''select sort_array(arr[10]),
+                  sort_array((select last(arr) from {table_name})[10])
+        from {table_name}
+        where (select first(arr) from {table_name})[0][1] > arr[0][1]
+        ''',
         conf=conf)
 
     # array of struct
@@ -120,10 +120,10 @@ def test_scalar_subquery_array(spark_tmp_table_factory, is_ansi_enabled, basic_g
         # Fix num_slices at 1 to make sure that first/last returns same results under CPU and GPU.
         lambda spark: gen_df(spark, [('arr', test_array_gen)], length=100, num_slices=1),
         table_name,
-        '''select arr[10].a, (select last(arr) from {})[10].a
-        from {}
-        where (select first(arr) from {})[0].a > arr[0].a
-        '''.format(table_name, table_name, table_name),
+        f'''select arr[10].a, (select last(arr) from {table_name})[10].a
+        from {table_name}
+        where (select first(arr) from {table_name})[0].a > arr[0].a
+        ''',
         conf=conf)
 
 
@@ -142,12 +142,12 @@ def test_scalar_subquery_array_ansi_mode_failures(spark_tmp_table_factory):
         # Fix num_slices at 1 to make sure that first/last returns same results under CPU and GPU.
         df = gen_df(spark, [('arr', ArrayGen(long_gen))], num_slices=1)
         df.createOrReplaceTempView(table_name)
-        query = '''
+        query = f'''
           SELECT SORT_ARRAY(arr),
-                 SORT_ARRAY((SELECT LAST(arr) FROM {}))
-          FROM {}
-          WHERE (SELECT FIRST(arr) FROM {})[0] > arr[0]
-        '''.format(table_name, table_name, table_name)
+                 SORT_ARRAY((SELECT LAST(arr) FROM {table_name}))
+          FROM {table_name}
+          WHERE (SELECT FIRST(arr) FROM {table_name})[0] > arr[0]
+        '''
         return spark.sql(query)
 
     assert_gpu_and_cpu_error(
@@ -168,30 +168,30 @@ def test_scalar_subquery_map(spark_tmp_table_factory):
         # Fix num_slices at 1 to make sure that first/last returns same results under CPU and GPU.
         lambda spark: gen_df(spark, [('kv', map_gen)], length=100, num_slices=1),
         table_name,
-        '''select kv['key_0'],
-                  (select first(kv) from {})['key_1'],
-                  (select last(kv) from {})['key_2']
-        from {}
-        '''.format(table_name, table_name, table_name))
+        f'''select kv['key_0'],
+                  (select first(kv) from {table_name})['key_1'],
+                  (select last(kv) from {table_name})['key_2']
+        from {table_name}
+        ''')
 
     # array of map
     assert_gpu_and_cpu_are_equal_sql(
         # Fix num_slices at 1 to make sure that first/last returns same results under CPU and GPU.
         lambda spark: gen_df(spark, [('arr', ArrayGen(map_gen, min_length=1))], length=100, num_slices=1),
         table_name,
-        '''select arr[0]['key_0'],
-                  (select first(arr) from {})[0]['key_1'],
-                  (select last(arr[0]) from {})['key_2']
-        from {}
-        '''.format(table_name, table_name, table_name))
+        f'''select arr[0]['key_0'],
+                  (select first(arr) from {table_name})[0]['key_1'],
+                  (select last(arr[0]) from {table_name})['key_2']
+        from {table_name}
+        ''')
 
     # struct of map
     assert_gpu_and_cpu_are_equal_sql(
         # Fix num_slices at 1 to make sure that first/last returns same results under CPU and GPU.
         lambda spark: gen_df(spark, [('ss', StructGen([['kv', map_gen]]))], length=100, num_slices=1),
         table_name,
-        '''select ss['kv']['key_0'],
-                  (select first(ss) from {})['kv']['key_1'],
-                  (select last(ss.kv) from {})['key_2']
-        from {}
-        '''.format(table_name, table_name, table_name))
+        f'''select ss['kv']['key_0'],
+                  (select first(ss) from {table_name})['kv']['key_1'],
+                  (select last(ss.kv) from {table_name})['key_2']
+        from {table_name}
+        ''')

--- a/integration_tests/src/main/python/subquery_test.py
+++ b/integration_tests/src/main/python/subquery_test.py
@@ -29,6 +29,7 @@ def test_scalar_subquery_basics(data_gen):
         where a > (select first(a) from table)
         ''')
 
+
 @ignore_order(local=True)
 @pytest.mark.parametrize('basic_gen', all_basic_gens, ids=idfn)
 def test_scalar_subquery_struct(basic_gen):
@@ -53,7 +54,11 @@ def test_scalar_subquery_struct(basic_gen):
         where (select first(ss) from table)['child']['c0'] > ss.child.c0
         ''')
     # struct of array
-    gen = [('ss', StructGen([['arr', ArrayGen(basic_gen)]]))]
+    # Note: The test query accesses the first two elements of the array.  The datagen is set up
+    #       to generate arrays of a minimum of two elements.  Otherwise, the test will fail in ANSI mode.
+    #       No meaningful test coverage is lost.  Accessing invalid indices of arrays is already tested
+    #       as part of array_test.py::test_array_item_ansi_fail_invalid_index.
+    gen = [('ss', StructGen([['arr', ArrayGen(basic_gen, min_length=2)]]))]
     assert_gpu_and_cpu_are_equal_sql(
         # Fix num_slices at 1 to make sure that first/last returns same results under CPU and GPU.
         lambda spark: gen_df(spark, gen, length=100, num_slices=1),
@@ -63,13 +68,18 @@ def test_scalar_subquery_struct(basic_gen):
         where (select first(ss) from table).arr[0] > ss.arr[1]
         ''')
 
+
 @ignore_order(local=True)
 @pytest.mark.parametrize('basic_gen', all_basic_gens, ids=idfn)
 def test_scalar_subquery_array(basic_gen):
+    # Note: For this test, all the array inputs are sized so that ArrayIndexOutOfBounds conditions are
+    #       avoided.  This is to ensure that the tests don't fail with exceptions in ANSI mode.
+    #       Note that no meaningful test coverage is lost here.  ArrayIndexOutOfBounds exceptions are
+    #       already tested as part of array_test.py::test_array_item_ansi_fail_invalid_index.
     # single-level array
     assert_gpu_and_cpu_are_equal_sql(
         # Fix num_slices at 1 to make sure that first/last returns same results under CPU and GPU.
-        lambda spark: gen_df(spark, [('arr', ArrayGen(basic_gen))], num_slices=1),
+        lambda spark: gen_df(spark, [('arr', ArrayGen(basic_gen, min_length=1))], num_slices=1),
         'table',
         '''select sort_array(arr),
                   sort_array((select last(arr) from table))
@@ -79,7 +89,7 @@ def test_scalar_subquery_array(basic_gen):
     # nested array
     assert_gpu_and_cpu_are_equal_sql(
         # Fix num_slices at 1 to make sure that first/last returns same results under CPU and GPU.
-        lambda spark: gen_df(spark, [('arr', ArrayGen(ArrayGen(basic_gen)))]
+        lambda spark: gen_df(spark, [('arr', ArrayGen(ArrayGen(basic_gen, min_length=2), min_length=11))]
                              , length=100
                              , num_slices=1),
         'table',
@@ -91,7 +101,7 @@ def test_scalar_subquery_array(basic_gen):
     # array of struct
     assert_gpu_and_cpu_are_equal_sql(
         # Fix num_slices at 1 to make sure that first/last returns same results under CPU and GPU.
-        lambda spark: gen_df(spark, [('arr', ArrayGen(StructGen([['a', basic_gen]])))]
+        lambda spark: gen_df(spark, [('arr', ArrayGen(StructGen([['a', basic_gen]]), min_length=11))]
                              , length=100
                              , num_slices=1),
         'table',
@@ -102,6 +112,10 @@ def test_scalar_subquery_array(basic_gen):
 
 @ignore_order(local=True)
 def test_scalar_subquery_map():
+    # Note: For this test, all the array inputs are sized so that ArrayIndexOutOfBounds conditions are
+    #       avoided.  This is to ensure that the tests don't fail with exceptions in ANSI mode.
+    #       Note that no meaningful test coverage is lost here.  ArrayIndexOutOfBounds exceptions are
+    #       already tested as part of array_test.py::test_array_item_ansi_fail_invalid_index.
     map_gen = map_string_string_gen[0]
     assert_gpu_and_cpu_are_equal_sql(
         # Fix num_slices at 1 to make sure that first/last returns same results under CPU and GPU.
@@ -115,7 +129,7 @@ def test_scalar_subquery_map():
     # array of map
     assert_gpu_and_cpu_are_equal_sql(
         # Fix num_slices at 1 to make sure that first/last returns same results under CPU and GPU.
-        lambda spark: gen_df(spark, [('arr', ArrayGen(map_gen))], length=100, num_slices=1),
+        lambda spark: gen_df(spark, [('arr', ArrayGen(map_gen, min_length=1))], length=100, num_slices=1),
         'table',
         '''select arr[0]['key_0'],
                   (select first(arr) from table)[0]['key_1'],

--- a/integration_tests/src/main/python/subquery_test.py
+++ b/integration_tests/src/main/python/subquery_test.py
@@ -146,7 +146,7 @@ def test_scalar_subquery_array_ansi_mode_failures(spark_tmp_table_factory):
           SELECT SORT_ARRAY(arr),
                  SORT_ARRAY((SELECT LAST(arr) FROM {table_name}))
           FROM {table_name}
-          WHERE (SELECT FIRST(arr) FROM {table_name})[0] > arr[0]
+          WHERE (SELECT CAST(ARRAY() AS ARRAY<BIGINT>))[0] > arr[0]
         '''
         return spark.sql(query)
 

--- a/integration_tests/src/main/python/subquery_test.py
+++ b/integration_tests/src/main/python/subquery_test.py
@@ -13,7 +13,7 @@
 # limitations under the License.
 
 import pytest
-from asserts import assert_gpu_and_cpu_are_equal_sql
+from asserts import assert_gpu_and_cpu_are_equal_collect, assert_gpu_and_cpu_are_equal_sql, assert_gpu_and_cpu_error
 from data_gen import *
 from marks import *
 
@@ -109,6 +109,44 @@ def test_scalar_subquery_array(basic_gen):
         from table
         where (select first(arr) from table)[0].a > arr[0].a
         ''')
+
+
+@ignore_order(local=True)
+@pytest.mark.parametrize('is_ansi_enabled', [False, True])
+def test_scalar_subquery_array_ansi_mode_failures(is_ansi_enabled):
+    """
+    This tests the case where the array scalar returned from a subquery might be indexed into
+    with an out-of-range index value:
+      - With ANSI mode enabled, an exception is expected.
+      - With ANSI mode disabled, a NULL value is expected.
+    This is only to test the corner case that used to be tested in `test_scalar_subquery_array`,
+    before ANSI support was added.
+    A more thorough test for invalid indices is done in array_test.py::test_array_item_ansi_fail_invalid_index,
+    and is out of the scope of this test.
+    """
+    conf = {'spark.sql.ansi.enabled': is_ansi_enabled}
+
+    def test_function(spark):
+        table_name = 'scalar_subquery_array_input'
+        # Fix num_slices at 1 to make sure that first/last returns same results under CPU and GPU.
+        df = gen_df(spark, [('arr', ArrayGen(long_gen))], num_slices=1)
+        df.createOrReplaceTempView(table_name)
+        query = '''
+          SELECT SORT_ARRAY(arr),
+                 SORT_ARRAY((SELECT LAST(arr) FROM {}))
+          FROM {}
+          WHERE (SELECT FIRST(arr) FROM {})[0] > arr[0]
+        '''.format(table_name, table_name, table_name)
+        return spark.sql(query)
+
+    if is_ansi_enabled:
+        assert_gpu_and_cpu_error(
+            lambda spark: test_function(spark).collect(),
+            conf=conf,
+            error_message='ArrayIndexOutOfBoundsException')
+    else:
+        assert_gpu_and_cpu_are_equal_collect(test_function, conf=conf)
+
 
 @ignore_order(local=True)
 def test_scalar_subquery_map():

--- a/integration_tests/src/main/python/subquery_test.py
+++ b/integration_tests/src/main/python/subquery_test.py
@@ -86,11 +86,11 @@ def test_scalar_subquery_array(is_ansi_enabled, basic_gen):
     assert_gpu_and_cpu_are_equal_sql(
         # Fix num_slices at 1 to make sure that first/last returns same results under CPU and GPU.
         lambda spark: gen_df(spark, [('arr', test_array_gen)], num_slices=1),
-        'table',
+        'my_array_table',
         '''select sort_array(arr),
-                  sort_array((select last(arr) from table))
-        from table
-        where (select first(arr) from table)[0] > arr[0]
+                  sort_array((select last(arr) from my_array_table))
+        from my_array_table
+        where (select first(arr) from my_array_table)[0] > arr[0]
         ''',
         conf=conf)
 

